### ISSUE TITLE
Skyframe cycle in RegisteredExecutionPlatformsFunction

### ISSUE THREAD

**Issue by katre** [bazelbuild/bazel#6831]
GRPC has reported a hard Bazel crash due to a cycle arising from RegisteredExecutionPlatformsFunction.

Command and error:
```
roth@segfault:~/grpc> bazel test test/cpp/end2end/client_lb_end2end_test --incompatible_remove_native_git_repository=false --incompatible_remove_native_http_archive=false
Starting local Bazel server and connecting to it...
INFO: Invocation ID: 0058df2b-dca2-4fd7-81e1-829d1eba8332
Internal error thrown during build. Printing stack trace: java.lang.IllegalStateException: topLevelKey: //test/cpp/end2end:client_lb_end2end_test BuildConfigurationValue.Key[5bf2e969a38eb49f752709c0acfde687] false
alreadyReported: false
path to cycle:
//test/cpp/end2end:client_lb_end2end_test BuildConfigurationValue.Key[5bf2e969a38eb49f752709c0acfde687] false
//:windows_msvc BuildConfigurationValue.Key[5bf2e969a38eb49f752709c0acfde687] false
cycle:
com.google.devtools.build.lib.skyframe.RegisteredExecutionPlatformsValue$Key@f6084b7a
//third_party/toolchains:cc-toolchain-clang-x86_64-default BuildConfigurationValue.Key[5bf2e969a38eb49f752709c0acfde687] false

        at com.google.common.base.Preconditions.checkState(Preconditions.java:507)
        at com.google.devtools.build.skyframe.CyclesReporter.reportCycles(CyclesReporter.java:85)
        at com.google.devtools.build.lib.skyframe.SkyframeBuildView.configureTargets(SkyframeBuildView.java:367)
        at com.google.devtools.build.lib.analysis.BuildView.update(BuildView.java:368)
        at com.google.devtools.build.lib.buildtool.AnalysisPhaseRunner.runAnalysisPhase(AnalysisPhaseRunner.java:212)
        at com.google.devtools.build.lib.buildtool.AnalysisPhaseRunner.execute(AnalysisPhaseRunner.java:121)
        at com.google.devtools.build.lib.buildtool.BuildTool.buildTargets(BuildTool.java:143)
        at com.google.devtools.build.lib.buildtool.BuildTool.processRequest(BuildTool.java:253)
        at com.google.devtools.build.lib.runtime.commands.TestCommand.doTest(TestCommand.java:131)
        at com.google.devtools.build.lib.runtime.commands.TestCommand.exec(TestCommand.java:111)
        at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:482)
        at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:204)
        at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:750)
        at com.google.devtools.build.lib.server.GrpcServerImpl.access$1600(GrpcServerImpl.java:103)
        at com.google.devtools.build.lib.server.GrpcServerImpl$2.lambda$run$0(GrpcServerImpl.java:819)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
```

**Comment by katre** [comment]
I was able to work around this by changeing grpc's WORKSPACE file to not use `:all` target patterns for `register_execution_platforms` and `register_toolchains`:
```
register_execution_platforms(
    "//third_party/toolchains:rbe_ubuntu1604",
    "//third_party/toolchains:rbe_ubuntu1604_large",
)

register_toolchains(
    "//third_party/toolchains:cc-toolchain-clang-x86_64-default",
)
```

However, we still need to fix this and make sure it cannot happen again.

**Comment by katre** [comment]
Email comment from @haxorz: is such a skyframe cycle possible? if so, SkyframeExecutor#createCyclesReporter needs to be updated?

**Comment by katre** [comment]
This appears to be fixed in Bazel HEAD, I'm trying to bisect back to see where that happened.

**Comment by katre** [comment]
Bisecting reveals that the issue is fixed by commit 402fd8bb9b505e555c39dc3c87c318453273cbca.

The next question to address is why.

**Comment by katre** [comment]
Okay, that commit fixes the problem by causing the `toolchain` rule to not require toolchain resolution. While this is the correct result, it's not a good general solution, because based on this, any use of `register_execution_platforms` with a wildcard target that expands to a target which _does_ require toolchain resolution will fail with this cycle error.

The target pattern expansion for `register_execution_platforms` needs to be fixed. Either:
1. Detect which targets require toolchains, and ignore them, or
2. Only allow target patterns to match targets which explicitly use the `platform` rule.

The second is annoying (it will prevent any future attempts to allow custom platform rules), but may be the best practical solution.
